stdlib/internal/promql/join.go

```diff
@@ -0,0 +1,554 @@
+	left, right execute.DatasetID
+	fn          *rowJoinFn
+
+	data *execute.GroupLookup
+	spec plan.TriggerSpec
+
+	ctx   context.Context
+	alloc *memory.Allocator
+}
+
+type cacheEntry struct {
+	l, r *RowIterator
+}
+
+func (e *cacheEntry) ready() bool {
+	return e.l != nil && e.r != nil && e.l.len != 0 && e.r.len != 0
+}
+
+func (c *mergeJoinCache) insert(id execute.DatasetID, key flux.GroupKey, iter *RowIterator) {
+	if entry, ok := c.data.Lookup(key); ok {
+		switch id {
+		case c.left:
+			entry.(*cacheEntry).l = iter
+		case c.right:
+			entry.(*cacheEntry).r = iter
+		}
+	} else {
+		switch id {
+		case c.left:
+			c.data.Set(key, &cacheEntry{l: iter})
+		case c.right:
+			c.data.Set(key, &cacheEntry{r: iter})
+		}
+	}
+}
+
+func (c *mergeJoinCache) delete(key flux.GroupKey) {
+	if entry, ok := c.data.Delete(key); ok {
+		t := entry.(*cacheEntry)
+		if t.l != nil {
+			for _, reader := range t.l.readers {
+				reader.Release()
+			}
+		}
+		if t.r != nil {
+			for _, reader := range t.r.readers {
+				reader.Release()
+			}
+		}
+	}
+}
+
+func (c *mergeJoinCache) clean() {
+	var keys []flux.GroupKey
+	c.data.Range(func(key flux.GroupKey, value interface{}) {
+		keys = append(keys, key)
+	})
+	for _, key := range keys {
+		c.delete(key)
+	}
+}
+
+func (c *mergeJoinCache) Table(key flux.GroupKey) (flux.Table, error) {
+	entry, ok := c.data.Lookup(key)
+	if !ok {
+		return nil, errors.Newf(codes.Internal, "no entry for group key %v in cache", key)
+	}
+	t := entry.(*cacheEntry)
+	if t.l == nil || t.r == nil {
+		return nil, errors.Newf(codes.Internal, "no entry for group key %v in cache", key)
+	}
+	return c.join(key, t.l, t.r)
+}
+
+func (c *mergeJoinCache) ForEach(f func(flux.GroupKey)) {
+	c.data.Range(func(key flux.GroupKey, value interface{}) {
+		if value.(*cacheEntry).ready() {
+			f(key)
+		}
+	})
+}
+
+func (c *mergeJoinCache) ForEachWithContext(f func(flux.GroupKey, execute.Trigger, execute.TableContext)) {
+	c.data.Range(func(key flux.GroupKey, value interface{}) {
+		if value.(*cacheEntry).ready() {
+			f(key, execute.NewTriggerFromSpec(c.spec), execute.TableContext{
+				Key: key,
+			})
+		}
+	})
+}
+
+func (c *mergeJoinCache) DiscardTable(key flux.GroupKey) {
+	c.delete(key)
+}
+
+func (c *mergeJoinCache) ExpireTable(key flux.GroupKey) {
+	c.delete(key)
+}
+
+func (c *mergeJoinCache) SetTriggerSpec(spec plan.TriggerSpec) {
+	c.spec = spec
+}
+
+func (c *mergeJoinCache) join(key flux.GroupKey, a, b *RowIterator) (flux.Table, error) {
+	// Compile row fn for the input rows
+	if err := c.fn.Prepare(a.columns(), b.columns()); err != nil {
+		return nil, err
+	}
+
+	// Create a table builder for the output of the join
+	builder := execute.NewColListTableBuilder(key, c.alloc)
+
+	firstRow := true
+	i, j := 0, 0
+
+NEXT:
+	ta := a.time(i)
+	tb := b.time(j)
+
+	if ta == -1 || tb == -1 {
+		goto DONE
+	}
+
+	if ta < tb {
+		i++
+		goto NEXT
+	}
+
+	if ta > tb {
+		j++
+		goto NEXT
+	}
+
+	for k := j; ta == b.time(k); k++ {
```

Comment:
What is this for loop doing? Can you add a comment to clarify? Specifically, it seems to me like the area above this has already verified that `ta == tb`. What additional logic does this add? I think what it is doing is allowing for the left side to be joined with multiple rows on the right side, but then I wonder why isn't the reverse possible. Shouldn't we be checking to see that if there are multiple times on the right side that join with times on the left side that we join those? Or do we want that behavior at all? Maybe we want to say that a join will not duplicate any rows.

I'd like a comment on this to clarify the intended behavior.